indra/media_plugins/cef/linux_volume_catcher_pw.cpp

```diff
@@ -0,0 +1,451 @@
+	spa_zero(mRegistryListener);
+	if (mCore) llpw_core_disconnect(mCore);
+	if (mContext) llpw_context_destroy(mContext);
+	pwUnlock();
+
+	if (!mThreadLoop) return;
+
+	llpw_thread_loop_stop(mThreadLoop);
+	llpw_thread_loop_destroy(mThreadLoop);
+
+	// debugPrint("cleanup done\n");
+}
+
+void VolumeCatcherImpl::pwLock() {
+	if (!mThreadLoop) return;
+	llpw_thread_loop_lock(mThreadLoop);
+}
+
+void VolumeCatcherImpl::pwUnlock() {
+	if (!mThreadLoop) return;
+	llpw_thread_loop_unlock(mThreadLoop);
+}
+
+// #define INV_LERP(a, b, v) (v - a) / (b - a)
+
+// #include <sys/time.h>
+
+void VolumeCatcherImpl::ChildNode::updateVolume()
+{
+	if (!mActive) return;
+
+	F32 volume = std::clamp(mImpl->mVolume, 0.0f, 1.0f);
+	// F32 pan = std::clamp(mImpl->mPan, -1.0f, 1.0f);
+
+	// debugClear();
+	// struct timeval time;
+	// gettimeofday(&time, NULL);
+	// double t = (double)time.tv_sec + (double)(time.tv_usec / 1000) / 1000;
+	// debugPrint("time is %f\n", t);
+	// F32 pan = std::sin(t * 2.0d);
+	// debugPrint("pan is %f\n", pan);
+
+	// uint32_t channels = 2;
+	// float volumes[channels];
+	// volumes[1] = INV_LERP(-1.0f, 0.0f, pan) * volume; // left
+	// volumes[0] = INV_LERP(1.0f, 0.0f, pan) * volume; // right
+
+	uint32_t channels = 1;
+	float volumes[channels];
+	volumes[0] = volume;
+
+	uint8_t buffer[512];
+
+	spa_pod_builder builder;
+	spa_pod_builder_init(&builder, buffer, sizeof(buffer));
+
+	spa_pod_frame frame;
+	spa_pod_builder_push_object(&builder, &frame, SPA_TYPE_OBJECT_Props, SPA_PARAM_Props);
+	spa_pod_builder_prop(&builder, SPA_PROP_channelVolumes, 0);
+	spa_pod_builder_array(&builder, sizeof(float), SPA_TYPE_Float, channels, volumes);
+	spa_pod* pod = static_cast<spa_pod*>(spa_pod_builder_pop(&builder, &frame));
+
+	mImpl->pwLock();
+	pw_node_set_param(mProxy, SPA_PARAM_Props, 0, pod);
+	mImpl->pwUnlock();
+}
+
+void VolumeCatcherImpl::ChildNode::destroy()
+{
+	if (!mActive) return;
+	mActive = false;
+
+	mImpl->mChildNodesMutex.lock();
+	mImpl->mChildNodes.erase(this);
+	mImpl->mChildNodesMutex.unlock();
+	
+	spa_hook_remove(&mNodeListener);
+	spa_hook_remove(&mProxyListener);
+
+	mImpl->pwLock();	
+	llpw_proxy_destroy(mProxy);
+	mImpl->pwUnlock();	
+}
+
+static pid_t getParentPid(pid_t aPid)
+{
+	std::stringstream strm;
+	strm << "/proc/" << aPid << "/status";
+	std::ifstream in{ strm.str() };
+
+	if (!in.is_open()) return 0;
+
+	pid_t res {0};
+	while (!in.eof() && res == 0)
+	{
+		std::string line;
+		line.resize( 1024, 0 );
+		in.getline( &line[0], line.length() );	
+
+		auto i = line.find("PPid:");
+		
+		if (i == std::string::npos) continue;
+		
+		char const *pIn = line.c_str() + 5; // Skip over pid;
+		while (*pIn != 0 && isspace( *pIn )) ++pIn;
+
+		if (*pIn) res = atoll(pIn);
+	}
+ 	return res;
+}
+
+static bool isPluginPid(pid_t aPid)
+{
+	auto myPid = getpid();
+
+	do
+	{
+		if(aPid == myPid) return true;
+		aPid = getParentPid( aPid );
+	} while(aPid > 1);
+
+	return false;
+}
+
```

Comment:
getParentPid and isPluginPid are function introduced in the pa volume catcher and duplicated here. The should be refactored into a common base (probably PluginMediaBase)